### PR TITLE
Close matplotlib figures after rendering them in `gr.Plot`

### LINKED ISSUE #11701: Clean figures passed to gr.Plot on Update
- [x] I have searched to see if a similar issue already exists.

**Is your feature request related to a problem? Please describe.**  
Figures naively (aka following the API guide and Tutorials) passed to gr.Plot will not be cleaned up if the Plot function is called again, leading to potential memory leaks or matplotlib not updating figures anymore.

A minimal example to reproduce this could be:

  ```
from typing import List,Tuple,Dict,Literal,Tuple 
  import os,sys 
  os.environ["GRADIO_TELEMETRY_ENABLED"] = "False" 
  
  import gradio as gr 
  
  import numpy as np 
  import matplotlib 
  import matplotlib.figure 
  import matplotlib.axes 
  import matplotlib.pyplot as plt
  
  
  def plot(slider_value:int):
      fig,ax1 = plt.subplots(1,1,figsize=(5,5))
      ax1.plot(np.arange(slider_value))
      ax1.set_title("arange N")
      return fig
  
  
  def main():
      with gr.Blocks() as demo:
          slider = gr.Slider(minimum=0,maximum=10,value=2,step=1,label="Plot Slider")
          show_plot = gr.Plot(label="Test Plot", format="png")
          slider.change(plot,inputs=[slider],outputs=[show_plot])
  
      demo.launch()

  if __name__ == "__main__":
      main()
```

 

**Describe the solution you'd like**  
The manual Solution to this is to create a State for each Figure, pass this to the function, check if its None or Not, and close old figures if present.

```
def plot2(slider_value:int,fig_cache:matplotlib.figure.Figure|None=None):
    if fig_cache is not None:
        plt.close(fig_cache)
    fig,ax1 = plt.subplots(1,1,figsize=(5,5))
    ax1.plot(np.arange(slider_value))
    ax1.set_title("arange N")
    return fig
```

 If this is in fact the intended way to do this, a little note in the Tutorials would be nice and might safe people some headache, if not, one might consider this a bug that needs fixing.

**Additional context**  
This is the Output one gets at the console after playing with the Slider for some time:

Image

### PR DESCRIPTION
**Pull request by hysts**
## Description

Matplotlib figures returned to `gr.Plot` were never closed after being encoded, so every event that updated a plot left another open figure in pyplot's global registry. Long-running apps accumulated figures, leaked memory, and eventually triggered matplotlib's "More than 20 figures have been opened" warning.

`Plot.postprocess` now calls `plt.close(value)` once the figure has been encoded to base64. Only `Figure` instances are closed. The legacy path where a handler returns the `pyplot` module itself is left untouched, since closing the current figure there could break apps that keep drawing on it across calls.

Closing a figure does not prevent it from being rendered again: `savefig` still works on a closed figure, so apps that hold on to a figure (for example in `gr.State`) and return it to `gr.Plot` repeatedly keep working. A test covers this, along with a regression test that the figure registry stays empty after `postprocess`. Figures created directly via `matplotlib.figure.Figure` (not through pyplot) are not in the registry, and closing them is a no-op.

Closes: #11701

## AI Disclosure

We encourage the use of AI tooling in creating PRs, but the any non-trivial use of AI needs be disclosed. E.g. if you used Claude to write a first draft, you should mention that. Trivial tab-completion doesn't need to be disclosed. You should self-review all PRs, especially if they were generated with AI.

- [x] I used AI to investigate the root cause and implement the fix.
- [ ] I did not use AI

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Testing and Formatting Your Code

1. PRs will only be merged if tests pass on CI. We recommend at least running the backend tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the backed tests: `bash scripts/run_backend_tests.sh`

2. Please run these bash scripts to automatically format your code: `bash scripts/format_backend.sh`, and (if you made any changes to non-Python files) `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/components/plot.py [hunk 1]
```diff
@@ -142,7 +143,17 @@ def postprocess(self, value: Any) -> PlotData | None:
             return value
         if isinstance(value, ModuleType) or "matplotlib" in value.__module__:
             dtype = "matplotlib"
-            out_y = processing_utils.encode_plot_to_base64(value, self.format)
+            try:
+                out_y = processing_utils.encode_plot_to_base64(value, self.format)
+            finally:
+                # A figure can only be in pyplot's registry if pyplot is
+                # already imported, so don't import it just to close.
+                plt = sys.modules.get("matplotlib.pyplot")
```
Comment (abidlabs) [inline gradio/components/plot.py]: In other parts of the codebase, we just do a try importing the module and then catch any importerror. I think that's more idiomatic than this.

File: test/components/test_plot.py [hunk 2]
```diff
@@ -58,6 +58,33 @@ def test_postprocess_altair(self):
...
+            plt.plot([1, 2, 3], [1, 2, 3])
+            component.postprocess(fig)
+            assert not plt.fignum_exists(fig.number)
+
+    def test_postprocess_accepts_closed_figure(self):
+        """
+        postprocess
+        """
+        with utils.MatplotlibBackendMananger():
+            import matplotlib.pyplot as plt
```
Comment (abidlabs) [inline test/components/test_plot.py]: move imports to top of file

**Review by abidlabs — APPROVED** [review]
Nice @hysts! LGTM, with a couple of style comments above

File: gradio/components/plot.py [hunk 3]
```diff
@@ -142,7 +142,17 @@ def postprocess(self, value: Any) -> PlotData | None:
...
             dtype = "matplotlib"
-            out_y = processing_utils.encode_plot_to_base64(value, self.format)
+            try:
+                out_y = processing_utils.encode_plot_to_base64(value, self.format)
+            finally:
+                try:
+                    import matplotlib.pyplot as plt
+                    from matplotlib.figure import Figure
+                except ImportError:
+                    pass
```
Comment (abidlabs) [inline gradio/components/plot.py]: Shouldn't pass here, it should return

Comment (abidlabs) [reply gradio/components/plot.py]: because this basically means that the module cannot be imported

**Comment by hysts** [comment]
Thanks for the review @abidlabs ! Addressed your comments. Merging.
